### ISSUE TITLE
Cisco Nexus switches and Routers failing

### ISSUE THREAD

**Issue by Jmandarino** [ytti/oxidized#644]
This is my first issue post so please forgive me if, My current issue is only with Nexus switches. When I try to run oxidized against them they fail with no error messages: below is an example:

`* Min threads: 0, max threads: 16
* Environment: development
* Listening on tcp://[iphere]
W, [2016-12-15T15:55:46.225320 #18964]  WARN -- : xxxxxx status fail, retry attempt 1
W, [2016-12-15T15:55:47.225985 #18964]  WARN -- : xxxxxx  status fail, retry attempt 2
W, [2016-12-15T15:55:48.226603 #18964]  WARN -- : xxxxxx  status fail, retry attempt 3
W, [2016-12-15T15:55:49.227038 #18964]  WARN -- : xxxxxx status fail, retries exhausted, giving up
`
With Debug mode:

```
W, [2016-12-15T15:58:45.343063 #19147]  WARN -- : xxxxxx  status fail, retry attempt 1
D, [2016-12-15T15:58:45.343148 #19147] DEBUG -- : lib/oxidized/worker.rb: Jobs 0, Want: 1
D, [2016-12-15T15:58:45.343289 #19147] DEBUG -- : lib/oxidized/worker.rb: Added xxxxxx  to the job queue
D, [2016-12-15T15:58:45.343322 #19147] DEBUG -- : lib/oxidized/worker.rb: 1 jobs running in parallel
D, [2016-12-15T15:58:45.343527 #19147] DEBUG -- : lib/oxidized/job.rb: Starting fetching process for xxxxxx  at 2016-12-15 20:58:45 UTC
D, [2016-12-15T15:58:45.343584 #19147] DEBUG -- : lib/oxidized/job.rb: Config fetched for xxxxxx  at 2016-12-15 20:58:45 UTC
W, [2016-12-15T15:58:46.343849 #19147]  WARN -- : xxxxxx  status fail, retry attempt 2
D, [2016-12-15T15:58:46.343930 #19147] DEBUG -- : lib/oxidized/worker.rb: Jobs 0, Want: 1
```

Has anyone else had this issue? Please let me know if you need me to include anything else

**Comment by ytti** [comment]
Can you include input debug:

```
input:
  debug: true
```

**Comment by Jmandarino** [comment]
@ytti  the output is exactly the same

**Comment by ytti** [comment]
You should have separate file by the device's name containing the session log.

**Comment by Jmandarino** [comment]
I have the configs sorted configs by groups, would this session log be in those folders? If so I do not have any for this particular switch.

**Comment by ytti** [comment]
it would be in the log directory

**Comment by Jmandarino** [comment]
I found what you are talking about. There are logs for other devices but none for the device in question.

**Comment by ytti** [comment]
And there are no crash files for them?

**Comment by Jmandarino** [comment]
correct, there are no crash files

**Comment by ytti** [comment]
any chance you can put sanitised version of failing node online reachable from 91.198.120.1?

**Comment by Jmandarino** [comment]
I definitely cannot put the node online to be reachable to anyone.

**Comment by laf** [comment]
You probably should at least put some more info up, we run a mix of 3/5 and 7k's with no issue in Oxidized.

**Comment by Jmandarino** [comment]
@laf Is there anything in particular that I could post that could help?

**Comment by ytti** [comment]
If you don't have box you can sanitise and put online how about shared ssh session (tmate) or screenshare via webex or equivalent? I don't know what I'm looking for, so it would be very time consuming to keep stabbing in the dark until someone bleeds.

**Comment by laf** [comment]
Or at least Oxidized version and the version of NX-OS on the Nexus devices? I can compare to our setup then at least.

**Comment by Jmandarino** [comment]
@laf I am using oxidized 0.18.0 the switches/routers that I am having trouble with are either NXOS: 7.0(3)I4(1) or  6.0(2)U3(7)

@ytti webex should work

**Comment by laf** [comment]
@Jmandarino We've got two devices with the 7.0 release you said and it works fine.

This is on oxidized-0.16.3 though.

**Comment by Jmandarino** [comment]
@laf I downgraded to 0.16.3 (however I installed the current version of oxidized-script/oxidized-web) and I am having the same issue as before.

**Comment by Jmandarino** [comment]
@ytti  I believe I have made progress on this issue. In NXOS.rb My version (0.19.0) (installed via gem) doesn't have telnet capabilities. I just looked over the most current version on github and it seems that someone added support. Do you think this could be the issue? When I try to copy and paste the updated version of nxos.rb on github to my server oxidized crashes. 


my version of nxos.rb:

`class NXOS < Oxidized::Model

  prompt /^(\r?[\w.@_()-]+[#]\s?)$/
  comment '! '

  cmd 'show version' do |cfg|
    cfg = cfg.each_line.take_while { |line| not line.match(/uptime/i) }
    comment cfg.join ""
  end

  cmd 'show inventory' do |cfg|
    comment cfg
  end

  cmd 'show running-config' do |cfg|
    cfg.gsub! /^!Time:[^\n]*\n/, ''
  end

  cfg :ssh  do
    post_login 'terminal length 0'
    pre_logout 'exit'
  end
end`

**Comment by ytti** [comment]
Do you have crash file?

**Comment by Jmandarino** [comment]
yes here it is:

`
2017-02-02 22:13:12 UTC
undefined method `reject' for nil:NilClass [NoMethodError]
--------------------------------------------------
/usr/local/lib/ruby/gems/2.1.0/gems/oxidized-0.19.0/lib/oxidized/model/ios.rb:45:in `block in <class:IOS>'
/usr/local/lib/ruby/gems/2.1.0/gems/oxidized-0.19.0/lib/oxidized/model/model.rb:94:in `instance_exec'
/usr/local/lib/ruby/gems/2.1.0/gems/oxidized-0.19.0/lib/oxidized/model/model.rb:94:in `cmd'
/usr/local/lib/ruby/gems/2.1.0/gems/oxidized-0.19.0/lib/oxidized/model/model.rb:136:in `block in get'
/usr/local/lib/ruby/gems/2.1.0/gems/oxidized-0.19.0/lib/oxidized/model/model.rb:135:in `each'
/usr/local/lib/ruby/gems/2.1.0/gems/oxidized-0.19.0/lib/oxidized/model/model.rb:135:in `get'
/usr/local/lib/ruby/gems/2.1.0/gems/oxidized-0.19.0/lib/oxidized/input/cli.rb:14:in `get'
/usr/local/lib/ruby/gems/2.1.0/gems/oxidized-0.19.0/lib/oxidized/node.rb:61:in `run_input'
/usr/local/lib/ruby/gems/2.1.0/gems/oxidized-0.19.0/lib/oxidized/node.rb:38:in `block in run'
/usr/local/lib/ruby/gems/2.1.0/gems/oxidized-0.19.0/lib/oxidized/node.rb:33:in `each'
/usr/local/lib/ruby/gems/2.1.0/gems/oxidized-0.19.0/lib/oxidized/node.rb:33:in `run'
/usr/local/lib/ruby/gems/2.1.0/gems/oxidized-0.19.0/lib/oxidized/job.rb:9:in `block in initialize'
`

However, I believe I moved past this issue. Oxidized now runs fine but still can't connect to those specific instances. I am getting this now when trying to connect:

` xxxxxxxx status no_connection, retry attempt 1
W, [2017-02-02T16:27:29.166921 #7920]  WARN -- : xxxxxxxx raised Timeout::Error with msg "timed out while waiting for more data"
`

It seems I'm now timing out instead of crashing when I try to telnet to a nxos device.

**Comment by ytti** [comment]
It looks like something is not producing output at all, and model tries to remove strings from non-existing output. 
If you can put sanitised device online, I'm sure I can fix it.

**Comment by Jmandarino** [comment]
@ytti is there any other solution besides putting a device online, there is no way for me to do that unfortunately.

**Comment by ytti** [comment]
Look at this code:

https://github.com/ytti/oxidized/blob/0.19.0/lib/oxidized/model/ios.rb#L43-L45

  1. we run ```show run ```
  2. we remove first 3 lines
  3. we remove NTP from remaining lines

3. assumes we're operating on a ```String```, in your case we are not, we are in-fact operating on a ```NilClass```, which does not have ```reject```method.

Now the problem is, why do we return instance of ```NilClass```at step2. The logical explanation is, we don't have 3 lines. Observe:

```
[1] pry(main)> "line1\nline2\nline3\nline4".each_line.to_a[3..-1]
=> ["line4"]
[2] pry(main)> "line1\nline2\n".each_line.to_a[3..-1]
=> nil
```

So likely your ```show run``` imply isn't producing sufficient amount of output. You should see this in input debug:

```
input:
  debug: true
```

**Comment by Jmandarino** [comment]
@ytti  I seriously appreciate all your help so far! Your explanation is perfect and I fully understand what you are saying. For these routers with issues, I am using the nxos.rb file which doesn't seem to strip the first three lines. Trying to follow oxidized in general, nxos.rb doesn't call ios.rb does it?

I am also running debug mode. 

the current error is (I am no longer getting NilClass):

`W, [2017-02-06T15:57:12.666234 #5573]  WARN -- : XXXXXXXXX raised Timeout::E                                                                                    rror with msg "timed out while waiting for more data"`

which seems to be coming from:

https://github.com/ytti/oxidized/blob/0b5af4573c7832f7fb1eeeb39c5e8c7375936bfd/lib/oxidized/input/telnet.rb#L107

**Comment by ytti** [comment]
The traceback you have was for ```ios.rb```. If this is separate issue, please provide the input debug, so we can see the session log:

```
input:
  debug: true
```

**Comment by Jmandarino** [comment]
I believe the issue with the traceback is no longer an issue. However, it was part of the larger issue at hand which I described in my last comment. 

here is the log:



`Connected to XXXXXXXXXXX.

*******************************************************************
*                                                                 *
*     xxxxxxxxxxxxxxxxxx                              *
*                                                                 *
*     xxxxxxxxxxxxxxxxxxx                                      *
*                                                                 *
*     Model:  Nexus 3048                                          *
*     Serial: xxx                                         *
*                                                                 *
*     !!!!! WARNING !!!!!                                         *
*     Unauthorized access and use of this system is               *
*     not permitted and is strictly prohibited by                 *
*     security policies, regulations, state and federal laws.     *
*     UNAUTHORIZED USERS ARE SUBJECT TO CRIMINAL AND              *
*     CIVIL PENALTIES AS WELL AS COMPANY-INITIATED                *
*     DISCIPLINARY PROCEEDINGS.                                   *
*                                                                 *
*******************************************************************
login: xxxx
Password:
Bad terminal type: "network". Will assume vt100.
Cisco Nexus Operating System (NX-OS) Software
TAC support: http://www.cisco.com/tac
Copyright (c) 2002-2015, Cisco Systems, Inc. All rights reserved.
The copyrights to certain works contained in this software are
owned by other third parties and used and distributed under
license. Certain components of this software are licensed under
the GNU General Public License (GPL) version 2.0 or the GNU
Lesser General Public License (LGPL) Version 2.1. A copy of each
such license is available at
http://www.opensource.org/licenses/gpl-2.0.php and
http://www.opensource.org/licenses/lgpl-2.1.php
^Mxxxxxxxx# terminal length 0^M
% Permission denied for the role
^Mxxxxxxxx# show version^M
Cisco Nexus Operating System (NX-OS) Software
TAC support: http://www.cisco.com/tac
Documents: http://www.cisco.com/en/US/products/ps9372/tsd_products_support_serie
s_home.html
Copyright (c) 2002-2015, Cisco Systems, Inc. All rights reserved.
The copyrights to certain works contained herein are owned by
other third parties and are used and distributed under license.
Some parts of this software are covered under the GNU Public
License. A copy of the license is available at
http://www.gnu.org/licenses/gpl.html.


Software
  BIOS:      version 1.4.0
  loader:    version N/A
  kickstart: version 6.0(2)U3(7)
  system:    version 6.0(2)U3(7)
  Power Sequencer Firmware:
             Module 1: version v4.4
  BIOS compile time:       12/09/2013
  kickstart image file is: bootflash:///n3000-uk9-kickstart.6.0.2.U3.7.bin
  kickstart compile time:  2/19/2015 2:00:00 [02/19/2015 04:44:40]
  system image file is:    bootflash:///n3000-uk9.6.0.2.U3.7.bin
  system compile time:     2/19/2015 2:00:00 [02/19/2015 09:59:10]
^[[7m--More--^[[m

`

it ends abruptly.

**Comment by ytti** [comment]
I propose you read the output.

**Comment by ytti** [comment]
Your policy does not allow turning pager off, thus pager prompt is presented, and pager prompt is not handled by the model.

**Comment by Jmandarino** [comment]
@ytti yes I see this! I am trying to understand your code, can you point me to the location where oxidized runs commands. 

I think the solution would be to first run in privileged mode, then turn off the pager. currently on login in oxidized attempts to turn the pager off. 

https://github.com/ytti/oxidized/blob/0.19.0/lib/oxidized/model/nxos.rb#L19-L22

**Comment by ytti** [comment]
What do you mean 'privileged mode', do you mean enable? The model does not have enable support, you can review other models how to implement enable.

Personally I think enable is entirely useless, and users should be just given what ever permissions they need. 
However, I will accept enable support as PR.

**Comment by Jmandarino** [comment]
yes exactly enable mode. I will do that thank you. 

I am going to update this thread and close it if I fix the issue so others can learn from it if need be.

**Comment by Jmandarino** [comment]
Okay full solution and update. First off thank you @ytti  for all your help.

I was having issues connecting to nexus switches specifically. I ran into 2 issues, the first being the nxos, didn't support telnet, the second being nxos didn't support enabled or privileged mode. 

on the current version, 0.19.0 I was able to fix this by modifying the nxos.rb file, mine currently uses the enable password and allows for telnet. 

`class NXOS < Oxidized::Model

  prompt /^(\r?[\w.@_()-]+[#]\s?)$/
  comment '! '

  cmd 'show version' do |cfg|
    cfg = cfg.each_line.take_while { |line| not line.match(/uptime/i) }
    comment cfg.join ""
  end

  cmd 'show inventory' do |cfg|
    comment cfg
  end

  cmd 'show running-config' do |cfg|
    cfg.gsub! /^!Time:[^\n]*\n/, ''
  end

  cfg :telnet, :ssh do
    # preferred way to handle additional passwords
    if vars :enable
      post_login do
        send "enable\n"
        cmd vars(:enable)
      end
    end
    post_login 'terminal length 0'
    post_login 'terminal width 0'
    pre_logout 'exit'
  end

  cfg :telnet do
    username /^login:/
    password /^Password:/
  end
end`

**Comment by ytti** [comment]
Please send pull request.
